drivers/platform/xilinx/uart.c

```diff
@@ -0,0 +1,338 @@
+	struct irq_desc	*irq_desc = xil_uart_desc->irq_desc;
+
+	if (xil_uart_desc->bytes_reveived > 0) {
+		ret = irq_source_disable(irq_desc, xil_uart_desc->irq_id);
+		if (ret < 0)
+			return ret;
+		ret = fifo_insert(&xil_uart_desc->fifo, xil_uart_desc->buff,
+				  xil_uart_desc->bytes_reveived);
+		if (ret < 0)
+			return ret;
+		xil_uart_desc->bytes_reveived = 0;
+		XUartPs_Recv(xil_uart_desc->instance, (u8*)(xil_uart_desc->buff),
+			     UART_BUFF_LENGTH);
+		ret = irq_source_enable(irq_desc, xil_uart_desc->irq_id);
+		if (ret < 0)
+			return ret;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Read byte from fifo.
+ * @param desc - Instance descriptor containing a fifo element
+ * @param data - read value.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+static int32_t uart_read_byte(struct uart_desc *desc, uint8_t *data)
+{
+	struct xil_uart_desc *xil_uart_desc = desc->extra;
+	int32_t ret;
+
+	while (xil_uart_desc->fifo == NULL) {
+		/* nothing in fifo, wait until something is received */
+		ret = uart_fifo_insert(desc);
+		if (ret < 0)
+			return ret;
+	}
+
+	*data = xil_uart_desc->fifo->data[xil_uart_desc->fifo_read_offset];
+	xil_uart_desc->fifo_read_offset++;
+
+	if (xil_uart_desc->fifo->len - xil_uart_desc->fifo_read_offset <= 0) {
+		xil_uart_desc->fifo_read_offset = 0;
+		xil_uart_desc->fifo = fifo_remove(xil_uart_desc->fifo);
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Read data from UART device.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t uart_read(struct uart_desc *desc, uint8_t *data, uint32_t bytes_number)
+{
+	ssize_t ret;
+	for (uint32_t i = 0; i < bytes_number; i++) {
+		ret = uart_read_byte(desc, &data[i]);
+		if (ret < 0)
+			return ret;
+	}
+
+	return bytes_number;
+}
+
+/**
+ * @brief Write data to UART device.
+ * @param desc - Instance of UART.
+ * @param data - Pointer to buffer containing data.
+ * @param bytes_number - Number of bytes to read.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t uart_write(struct uart_desc *desc, const uint8_t *data,
+		   uint32_t bytes_number)
+{
+	struct xil_uart_desc *xil_uart_desc = desc->extra;
+	size_t total_sent = XUartPs_Send(xil_uart_desc->instance, (u8*)data,
+					 bytes_number);
+	while (XUartPs_IsSending(xil_uart_desc->instance));
+	if (total_sent < bytes_number)
+		return FAILURE;
+
+	return SUCCESS;
+}
+
+/**
+ * @brief UART interrupt handler.
+ * @param call_back_ref - Instance of UART.
+ * @param event - Event that caused the interrupt.
+ * @param data_len - number of bytes received.
+ */
+static void uart_irq_handler(void *call_back_ref, uint32_t event,
+			     uint32_t data_len)
+{
+	struct xil_uart_desc *xil_uart_desc = call_back_ref;
+	/* All of the data has been received */
+	if (event == XUARTPS_EVENT_RECV_DATA) {
```

Comment:
Would this string of `if`'s be better as a `switch` clause?

Reply:
Probably, I will change that.